### ISSUE TITLE
Flooded with "Two custom restoreOriginalUri callbacks are detected" warnings

### ISSUE THREAD

**Issue by indignant** [okta/okta-react#170]
## I'm submitting a:

- [x] Bug report
- [ ] Feature request
- [ ] Other (Describe below)

## Current behavior
Hello 🌈 I'm having some issues with [this line](https://github.com/okta/okta-react/blob/9bd630511d8ec6d071c6cb160457bf96b5d4d3a5/src/Security.tsx#L52) flooding my browser console and test outputs with this message:

```
Two custom restoreOriginalUri callbacks are detected.
The one from the OktaAuth configuration will be overridden
by the provided restoreOriginalUri prop from the Security component.
```

The code works as expected, but it's just a very noisy message. For instance, just picking a random test, it logs the message and the stack trace 13 times from one test file. So more useful messages are getting buried.

## Expected behavior
Do we really need this message at all? The code is working as expected, so it seems like it's just noise.

## Minimal reproduction of the problem with instructions
I can't share my code, but I can try to figure out a reproduction if no one else is experiencing it. Here's some of what I'm doing:

``` JS
export const oktaAuth = new OktaAuth({
  issuer: `https://${process.env.REACT_APP_OKTA_DOMAIN}/oauth2/default`,
  clientId: process.env.REACT_APP_OKTA_CLIENT_ID,
  redirectUri: window.location.origin + '/login/callback',
})

function SecurityWrapper({ children }) {
  const history = useHistory()
  async function restoreOriginalUri(_oktaAuth, originalUri) {
    history.replace(toRelativeUrl(originalUri || '/', window.location.origin))
  }

  return (
    <Security oktaAuth={oktaAuth}
      restoreOriginalUri={restoreOriginalUri}>
      {children}
    </Security>
  )
}
```

This is the only place the word `restoreOriginalUri` appears in my app. I only export `oktaAuth` for setting axios token and logging out.

With `useCallback` I can quiet the browser console, but my tests are still whining.

``` JS
export const oktaAuth = new OktaAuth({
  issuer: `https://${process.env.REACT_APP_OKTA_DOMAIN}/oauth2/default`,
  clientId: process.env.REACT_APP_OKTA_CLIENT_ID,
  redirectUri: window.location.origin + '/login/callback',
})

function SecurityWrapper({ children }) {
  const history = useHistory()
  const restoreOriginalUri = useCallback(
    async function(_oktaAuth, originalUri) {
      history.replace(toRelativeUrl(originalUri || '/', window.location.origin))
    },
    [history],
  )

  return (
    <Security oktaAuth={oktaAuth}
      restoreOriginalUri={restoreOriginalUri}>
      {children}
    </Security>
  )
}
```

## Extra information about the use case/user story you are trying to implement
I'm following react-testing-library's advice on [creating a custom render to wrap components in providers](https://testing-library.com/docs/react-testing-library/setup#custom-render).

## Environment

- Package version: `^6.2.0` for okta-react and `^5.5.0` for okta-auth-js
- React version: `^17.0.0`
- Browser: Firefox
- OS: OSX
- Node version (`node -v`): `v16.0.0`
- Other:

## Configuration

- Okta application type: SPA

Thanks 🦄

**Comment by DingessXL** [comment]
I had a similar problem but was able to get rid of the message by moving the initialization of OktaAuth into the the same component where the Security component was being used.

Paraphrasing some code here, as I can't post my source, this caused an issue:

```TSX
const oktaAuth = new OktaAuth({
  issuer: `https://${process.env.REACT_APP_OKTA_DOMAIN}/oauth2/default`,
  clientId: process.env.REACT_APP_OKTA_CLIENT_ID,
  redirectUri: window.location.origin + '/login/callback',
})

const App: React.FC = () => {
  const restoreOriginalUri = async (_oktaAuth: OktaAuth, originalUri: string) => {
    history.replace(toRelativeUrl(originalUri, window.location.origin));
  };
return (  <Security oktaAuth={oktaAuth} restoreOriginalUri={restoreOriginalUri} onAuthRequired={customAuthHandler}></Security>
}

```

But this didnt:

```TSX
const App: React.FC = () => {
const oktaAuth = new OktaAuth({
  issuer: `https://${process.env.REACT_APP_OKTA_DOMAIN}/oauth2/default`,
  clientId: process.env.REACT_APP_OKTA_CLIENT_ID,
  redirectUri: window.location.origin + '/login/callback',
})
  const restoreOriginalUri = async (_oktaAuth: OktaAuth, originalUri: string) => {
    history.replace(toRelativeUrl(originalUri, window.location.origin));
  };
return (  <Security oktaAuth={oktaAuth} restoreOriginalUri={restoreOriginalUri} onAuthRequired={customAuthHandler}></Security>
}

```

**Comment by shuowu** [comment]
OktaAuth initialization should be isolated from the component rendering logic, so only one OktaAuth instance is used in the React SPA.

**Comment by chinanderm** [comment]
There should be no reason why initializing a new `OktaAuth` should have to be done in the same component as where the `Security` component is used.

**Comment by chinanderm** [comment]
I too have my `OktaAuth` init outside of the component that uses `Security` and am also getting flooded with the message. What has fixed it for me is memoizing the `restoreOriginalUrl` function:

```typescript
const restoreOriginalUri = useCallback(async (_: unknown, originalUri: string) => {
  history.replace(toRelativeUrl(originalUri || '/', window.location.origin))
  // eslint-disable-next-line react-hooks/exhaustive-deps
}, [])

return (
  <Security
    oktaAuth={oktaAuth}
    onAuthRequired={routeToLoginPage}
    restoreOriginalUri={restoreOriginalUri}
  >
    ...
  </Security>
)
```

**Comment by Ziao** [comment]
Oh Okta. Your developer experience is awful. Half the time I have issues in my webapp it's because of this library. I don't even know what to say but isn't it time to put more capable people on this?

**Comment by rosedo** [comment]
The compiled code is invalid:
```
if (oktaAuth.options.restoreOriginalUri) {
      console.warn('Two custom restoreOriginalUri callbacks are detected. The one from the OktaAuth configuration will be overridden by the provided restoreOriginalUri prop from the Security component.');
    }
    oktaAuth.options.restoreOriginalUri = /*#__PURE__*/function () {
```

Correct me if I'm wrong but it seems that here, the way javascript works is that the oktaAuth.options.restoreOriginalUri will be defined earlier than where it's written because it uses the function keyword

**Comment by kashaf-s** [comment]
> I had a similar problem but was able to get rid of the message by moving the initialization of OktaAuth into the the same component where the Security component was being used.
> 
> Paraphrasing some code here, as I can't post my source, this caused an issue:
> 
> ```tsx
> const oktaAuth = new OktaAuth({
>   issuer: `https://${process.env.REACT_APP_OKTA_DOMAIN}/oauth2/default`,
>   clientId: process.env.REACT_APP_OKTA_CLIENT_ID,
>   redirectUri: window.location.origin + '/login/callback',
> })
> 
> const App: React.FC = () => {
>   const restoreOriginalUri = async (_oktaAuth: OktaAuth, originalUri: string) => {
>     history.replace(toRelativeUrl(originalUri, window.location.origin));
>   };
> return (  <Security oktaAuth={oktaAuth} restoreOriginalUri={restoreOriginalUri} onAuthRequired={customAuthHandler}></Security>
> }
> ```
> 
> But this didnt:
> 
> ```tsx
> const App: React.FC = () => {
> const oktaAuth = new OktaAuth({
>   issuer: `https://${process.env.REACT_APP_OKTA_DOMAIN}/oauth2/default`,
>   clientId: process.env.REACT_APP_OKTA_CLIENT_ID,
>   redirectUri: window.location.origin + '/login/callback',
> })
>   const restoreOriginalUri = async (_oktaAuth: OktaAuth, originalUri: string) => {
>     history.replace(toRelativeUrl(originalUri, window.location.origin));
>   };
> return (  <Security oktaAuth={oktaAuth} restoreOriginalUri={restoreOriginalUri} onAuthRequired={customAuthHandler}></Security>
> }
> ```

I have the same exact issue, but this solution isn't working for me. Any idea why?

**Comment by kashaf-s** [comment]
> @kashaf-s What version are you using? This issue was addressed here: #239

 "@okta/okta-auth-js": "^6.0.0",
    "@okta/okta-react": "^6.4.3",
    "react-router-dom": "^5.3.4",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-scripts": "5.0.1"

these are my package versions. Can you tell me which ones I should update to?

**Comment by jaredperreault-okta** [comment]
@kashaf-s The fix was released in `okta-react@6.6.0`

**Comment by kashaf-s** [comment]
> @kashaf-s The fix was released in `okta-react@6.6.0`

I upgraded and it's still not worked for me :(

**Comment by eliprand** [comment]
So, I can reproduce on the stock example for [react router v6](https://github.com/okta/okta-react/tree/master/samples/routing/react-router-dom-v6), upgraded to React 18.
If you simply change how the app is bootstrapped and use `ReactDOM.createRoot()`, it starts throwing the warning.
